File: python/4-median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        def joinSortedLists(list1, list2):
            p1 = p2 = 0
            out = []
            while p1 < len(list1) or p2 < len(list2):
                if p1 == len(list1):
                    out.append(list2[p2])
                    p2 += 1
                elif p2 == len(list2) or list1[p1] < list2[p2]:
                    out.append(list1[p1])
                    p1 += 1
                else:
                    out.append(list2[p2])
                    p2 += 1
            return out
        
        def arrayMedian(a):
            if len(a) % 2 == 0:
                return (a[len(a)//2 - 1] + a[len(a)//2]) / 2
            return a[len(a)//2]
        
        joined = joinSortedLists(nums1, nums2)
        result = arrayMedian(joined)
        return result

File: java/4-median-of-two-sorted-arrays.java
class Solution {

    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int[] joined = joinSortedLists(nums1, nums2);
        return arrayMedian(joined);
    }

    public int[] joinSortedLists(int[] list1, int[] list2) {
        int p1 = 0, p2 = 0, ptr = 0;
        int[] out = new int[list1.length + list2.length];
        while (p1 < list1.length || p2 < list2.length) {
            if (p1 == list1.length) {
                out[ptr++] = list2[p2++];
            } else if (p2 == list2.length || list1[p1] < list2[p2]) {
                out[ptr++] = list1[p1++];
            } else {
                out[ptr++] = list2[p2++];
            }
        }
        return out;
    }

    public double arrayMedian(int[] a) {
        if (a.length % 2 == 0) {
            return (double) (a[a.length / 2 - 1] + a[a.length / 2]) / 2.0;
        }
        return a[a.length / 2];
    }

}
